prov/gni/src/gnix_msg.c

```diff
@@ -1904,31 +1952,15 @@ static int __smsg_eager_msg_w_data(void *data, void *msg)
 
 		__gnix_msg_copy_data_to_recv_addr(req, data_ptr);
 
-		if ((req->msg.recv_flags & FI_MULTI_RECV) &&
-		    ((req->msg.cum_recv_len - req->msg.cum_send_len) <
-			ep->min_multi_recv))
-			req->msg.recv_flags &= ~GNIX_MSG_MULTI_RECV_SUP;
-
 		__gnix_msg_recv_completion(ep, req);
 
-		/* Check if we're using FI_MULTI_RECV and there is space left
-		 * in the receive buffer. */
-		if ((req->msg.recv_flags & FI_MULTI_RECV) &&
-		    ((req->msg.cum_recv_len - req->msg.cum_send_len) >=
-			     ep->min_multi_recv)) {
-			GNIX_DEBUG(FI_LOG_EP_DATA, "Re-using req: %p\n", req);
-
-			/* Adjust receive buffer for the next match. */
-			req->msg.recv_info[0].recv_addr += req->msg.cum_send_len;
-			req->msg.recv_info[0].recv_len -= req->msg.cum_send_len;
-			req->msg.cum_recv_len = req->msg.recv_info[0].recv_len;
-		} else {
-			GNIX_DEBUG(FI_LOG_EP_DATA, "Freeing req: %p\n", req);
+		GNIX_DEBUG(FI_LOG_EP_DATA, "Freeing req: %p\n", req);
 
-			/* Dequeue and free the request. */
+		/* Dequeue and free the request. */
+		if (mrecv_req == NULL)
 			_gnix_remove_tag(posted_queue, req);
-			_gnix_fr_free(ep, req);
-		}
+		_gnix_fr_free(ep, req);
```

Comment:
i see the code is a little confusing.  I'll put both statements inside the 'if'.